include/oneapi/dpl/experimental/kt/single_pass_scan.h

```diff
@@ -0,0 +1,413 @@
+    ::std::size_t __n;
+    _StatusFlags __status_flags;
+    ::std::size_t __status_flags_size;
+    _StatusValues __status_vals_full;
+    _StatusValues __status_vals_partial;
+    ::std::size_t __current_num_items;
+    _TileVals __tile_vals;
+
+    [[sycl::reqd_sub_group_size(SUBGROUP_SIZE)]] void
+    operator()(const sycl::nd_item<1>& __item) const
+    {
+        auto __group = __item.get_group();
+        auto __subgroup = __item.get_sub_group();
+        auto __local_id = __item.get_local_id(0);
+
+        ::std::uint32_t __tile_id = 0;
+
+        // Obtain unique ID for this work-group that will be used in decoupled lookback
+        if (__group.leader())
+        {
+            sycl::atomic_ref<_FlagStorageType, sycl::memory_order::relaxed, sycl::memory_scope::device,
+                             sycl::access::address_space::global_space>
+                __idx_atomic(__status_flags[__status_flags_size - 1]);
+            __tile_id = __idx_atomic.fetch_add(1);
+        }
+
+        __tile_id = sycl::group_broadcast(__group, __tile_id, 0);
+
+        ::std::size_t __current_offset = static_cast<::std::size_t>(__tile_id) * __elems_in_tile;
+        auto __out_begin = __out_rng.begin() + __current_offset;
+
+        if (__current_offset >= __n)
+            return;
+
+        // Global load into local
+        auto __wg_current_offset = (__tile_id * __elems_in_tile);
+        auto __wg_next_offset = ((__tile_id + 1) * __elems_in_tile);
+        auto __wg_local_memory_size = __elems_in_tile;
+
+        if (__wg_next_offset > __n)
+            __wg_local_memory_size = __n - __wg_current_offset;
+
+        if (__wg_next_offset <= __n)
+        {
+            _ONEDPL_PRAGMA_UNROLL
+            for (std::uint32_t __i = 0; __i < __data_per_workitem; ++__i)
+            {
+                __tile_vals[__local_id + __workgroup_size * __i] =
+                    __in_rng[__wg_current_offset + __local_id + __workgroup_size * __i];
+            }
+        }
+        else
+        {
+            _ONEDPL_PRAGMA_UNROLL
+            for (std::uint32_t __i = 0; __i < __data_per_workitem; ++__i)
+            {
+                if (__wg_current_offset + __local_id + __workgroup_size * __i < __n)
+                {
+                    __tile_vals[__local_id + __workgroup_size * __i] =
+                        __in_rng[__wg_current_offset + __local_id + __workgroup_size * __i];
+                }
+            }
+        }
+
+        auto __tile_vals_ptr = __dpl_sycl::__get_accessor_ptr(__tile_vals);
+        _Type __local_reduction =
+            sycl::joint_reduce(__group, __tile_vals_ptr, __tile_vals_ptr + __wg_local_memory_size, __binary_op);
+        _Type __prev_tile_reduction{};
+
+        // The first sub-group will query the previous tiles to find a prefix
+        if (__subgroup.get_group_id() == 0)
+        {
+            _FlagType __flag(__status_flags, __status_vals_full, __status_vals_partial, __tile_id);
+
+            if (__group.leader())
```

Comment:
According to the SYCL 2020 spec, "The leader of the work-group is guaranteed to be the work-item with a local id of 0." However, from what I can tell, I do not find any guarantees that work-item 0 is guaranteed to be in the first subgroup. I find "work-items that compose a sub-group are chosen in an implementation defined way" which implies that it actually is not guaranteed.

Practically, it might be the case, but I will restructure this code so that it does not rely on something not explicitly guaranteed by the spec. I believe rewriting it from `__group.leader()` to `__subgroup.get_local_id() == 0` would suffice.

Reply:
Right, this is what I found as well.  I agree with your proposal.